--- ../../plenoxels/svox2/utils.py	2024-02-13 12:06:07.930166019 -0500
+++ ../google/plenoxels/svox2/utils.py	2024-02-13 10:46:24.460645148 -0500
@@ -32,13 +32,16 @@
         import svox2.csrc as _C
         if not hasattr(_C, "sample_grid"):
             _C = None
-    except:
+    except Exception as e:
         _C = None
+        print("Exception:")
+        print(e)
 
     if _C is None:
         warn("CUDA extension svox2.csrc could not be loaded! " +
              "Operations will be slow.\n" +
              "Please do not import svox in the svox2 source directory.")
+
     return _C
 
 
