--- external/stylegan/metrics/metric_main.py	2023-04-06 03:51:55.924849254 +0000
+++ external_reference/stylegan/metrics/metric_main.py	2023-04-06 03:41:03.682641908 +0000
@@ -89,6 +89,12 @@
     return dict(fid50k_full=fid)
 
 @register_metric
+def fid5k_full(opts):
+    opts.dataset_kwargs.update(max_size=None, xflip=False)
+    fid = frechet_inception_distance.compute_fid(opts, max_real=None, num_gen=5000)
+    return dict(fid5k_full=fid)
+
+@register_metric
 def kid50k_full(opts):
     opts.dataset_kwargs.update(max_size=None, xflip=False)
     kid = kernel_inception_distance.compute_kid(opts, max_real=1000000, num_gen=50000, num_subsets=100, max_subset_size=1000)
