--- ../../swvox/swvox/csrc/CMakeLists.txt	2024-02-14 05:16:57.537407540 -0500
+++ ../google/swvox/swvox/csrc/CMakeLists.txt	2024-02-13 10:46:24.584645005 -0500
@@ -40,7 +40,7 @@
     cmake_policy(SET CMP0072 NEW)
 endif (POLICY CMP0072)
 
-project( svox )
+project( swvox )
 
 enable_language(CUDA)
 message(STATUS "CUDA enabled")
@@ -64,13 +64,13 @@
 
 include_directories (${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
 
-pybind11_add_module(svox-test SHARED ${SOURCES})
-target_link_libraries(svox-test PRIVATE "${TORCH_LIBRARIES}")
-target_include_directories(svox-test PRIVATE "${INCLUDE_DIR}")
+pybind11_add_module(swvox-test SHARED ${SOURCES})
+target_link_libraries(swvox-test PRIVATE "${TORCH_LIBRARIES}")
+target_include_directories(swvox-test PRIVATE "${INCLUDE_DIR}")
 
 if (MSVC)
   file(GLOB TORCH_DLLS "${TORCH_INSTALL_PREFIX}/lib/*.dll")
-  add_custom_command(TARGET svox-test
+  add_custom_command(TARGET swvox-test
                      POST_BUILD
                      COMMAND ${CMAKE_COMMAND} -E copy_if_different
                      ${TORCH_DLLS}
