--- external/eg3d/training/dual_discriminator.py	2023-04-06 03:56:53.605414776 +0000
+++ external_reference/eg3d/training/dual_discriminator.py	2023-04-06 03:41:04.690731872 +0000
@@ -15,7 +15,7 @@
 import torch
 from torch_utils import persistence
 from torch_utils.ops import upfirdn2d
-from training.networks_stylegan2 import DiscriminatorBlock, MappingNetwork, DiscriminatorEpilogue
+from external.stylegan.training.networks_stylegan2_terrain import DiscriminatorBlock, MappingNetwork, DiscriminatorEpilogue
 
 @persistence.persistent_class
 class SingleDiscriminator(torch.nn.Module):
