--- external/gsn/models/discriminator.py	2023-03-09 18:13:17.896724732 +0000
+++ external_reference/gsn/models/discriminator.py	2023-03-09 18:07:58.458065115 +0000
@@ -1,7 +1,9 @@
 import math
 import torch
 import torch.nn as nn
-from .layers import ConvLayer2d, ConvResBlock2d, EqualLinear
+from external.gsn.models.layers import ConvLayer2d, ConvResBlock2d, EqualLinear
+
+from torch_utils import persistence
 
 
 class DiscriminatorHead(nn.Module):
@@ -48,6 +50,7 @@
         return out
 
 
+@persistence.persistent_class
 class ConvDecoder(nn.Module):
     def __init__(self, in_channel, out_channel, in_res, out_res):
         super().__init__()
