--- ../../swvox/MANIFEST.in	2024-02-14 05:16:57.525407556 -0500
+++ ../google/swvox/MANIFEST.in	2024-02-13 10:46:24.568645024 -0500
@@ -2,16 +2,16 @@
 include LICENSE.txt
 include MANIFEST.in
 include setup.py
-include svox/svox.py
-include svox/helpers.py
-include svox/sh.py
-include svox/renderer.py
-include svox/__init__.py
-include svox/version.py
-include svox/csrc/include/common.cuh
-include svox/csrc/include/data_spec.hpp
-include svox/csrc/include/data_spec_packed.cuh
-include svox/csrc/svox.cpp
-include svox/csrc/svox_kernel.cu
-include svox/csrc/rt_kernel.cu
-include svox/csrc/quantizer.cpp
+include swvox/swvox.py
+include swvox/helpers.py
+include swvox/sh.py
+include swvox/renderer.py
+include swvox/__init__.py
+include swvox/version.py
+include swvox/csrc/include/common.cuh
+include swvox/csrc/include/data_spec.hpp
+include swvox/csrc/include/data_spec_packed.cuh
+include swvox/csrc/swvox.cpp
+include swvox/csrc/swvox_kernel.cu
+include swvox/csrc/rt_kernel.cu
+include swvox/csrc/quantizer.cpp
