--- ../../swvox/setup.py	2024-02-14 05:16:57.537407540 -0500
+++ ../google/swvox/setup.py	2024-02-13 10:46:24.572645020 -0500
@@ -6,19 +6,19 @@
 ROOT_DIR = osp.dirname(osp.abspath(__file__))
 
 __version__ = None
-exec(open('svox/version.py', 'r').read())
+exec(open('swvox/version.py', 'r').read())
 
 CUDA_FLAGS = []
 INSTALL_REQUIREMENTS = []
 
 try:
     ext_modules = [
-        CUDAExtension('svox.csrc', [
-            'svox/csrc/svox.cpp',
-            'svox/csrc/svox_kernel.cu',
-            'svox/csrc/rt_kernel.cu',
-            'svox/csrc/quantizer.cpp',
-        ], include_dirs=[osp.join(ROOT_DIR, "svox", "csrc", "include")],
+        CUDAExtension('swvox.csrc', [
+            'swvox/csrc/swvox.cpp',
+            'swvox/csrc/swvox_kernel.cu',
+            'swvox/csrc/rt_kernel.cu',
+            'swvox/csrc/quantizer.cpp',
+        ], include_dirs=[osp.join(ROOT_DIR, "swvox", "csrc", "include")],
         optional=True),
     ]
 except:
@@ -27,7 +27,7 @@
     ext_modules = []
 
 setup(
-    name='svox',
+    name='swvox',
     version=__version__,
     author='Alex Yu',
     author_email='alexyu99126@gmail.com',
@@ -35,7 +35,7 @@
     long_description='Sparse voxel N^3-tree data structure PyTorch extension, using CUDA',
     ext_modules=ext_modules,
     setup_requires=['pybind11>=2.5.0'],
-    packages=['svox', 'svox.csrc'],
+    packages=['swvox', 'swvox.csrc'],
     cmdclass={'build_ext': BuildExtension},
     zip_safe=False,
 )
