--- ../../plenoxels/svox2/csrc/render_svox1_kernel.cu	2024-02-13 12:06:07.930166019 -0500
+++ ../google/plenoxels/svox2/csrc/render_svox1_kernel.cu	2024-02-13 10:46:24.468645138 -0500
@@ -66,7 +66,7 @@
         if (ray.dir[i] == 0.f)
             invdir[i] = 1e9f;
     }
-
+    
     {
         float t1, t2;
         t = 0.0f;
@@ -247,9 +247,10 @@
                         if (tmp > 0.f) {
                             total_color += tmp * grad_output[k];
                             tmp = weight * grad_output[k];
+                            //printf("Substituted backward for just sh's for debugging, undo!!");
                             for (int i = 0; i < grid.basis_dim; ++i) {
-                                atomicAdd(&grad_sample_val[off + i],
-                                        basis_fn[i] * tmp);
+                                // correct line is next, others are just for debug:
+                                atomicAdd(&grad_sample_val[off + i], basis_fn[i] * tmp);
                             }
                         }
                     }
