--- ../../plenoctree/octree/nerf/utils.py	2024-02-13 12:39:49.535388381 -0500
+++ ../google/plenoctree/octree/nerf/utils.py	2024-02-13 10:46:24.364645257 -0500
@@ -16,6 +16,10 @@
 
 # Lint as: python3
 """Utility functions."""
+import sys
+sys.path.append('..')
+sys.path.append('../..')
+
 import collections
 import os
 from os import path
@@ -223,9 +227,12 @@
     """Update the flags in `args` with the contents of the config YAML file."""
     if args.config is None:
         return
-    pth = path.join(args.config + ".yaml")
+    if not args.config.endswith('.yaml'):
+        pth = path.join(args.config + ".yaml")
+    else:
+        pth = args.config
     with open_file(pth, "r") as fin:
-        configs = yaml.load(fin, Loader=yaml.FullLoader)
+        configs = yaml.safe_load(fin)
     # Only allow args to be updated if they already exist.
     invalid_args = list(set(configs.keys()) - set(dir(args)))
     if invalid_args:
