--- ../../plenoctree/octree/extraction.py	2024-02-13 12:39:49.531388387 -0500
+++ ../google/plenoctree/octree/extraction.py	2024-02-13 10:46:24.364645257 -0500
@@ -35,6 +35,9 @@
     --data_dir $DATA_ROOT/$SCENE/ \
     --output $CKPT_ROOT/$SCENE/octrees/tree.npz
 """
+import cv2
+import sys
+sys.path.append('plenoctree')
 import os
 # Get rid of ugly TF logs
 os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
@@ -56,6 +59,9 @@
 from svox import NDCConfig, VolumeRenderer
 from svox.helpers import _get_c_extension
 from tqdm import tqdm
+import sys
+sys.path.append('..')
+from my_python_utils.common_utils import *
 
 _C = _get_c_extension()
 
@@ -374,7 +380,7 @@
         else:  # vanilla NeRF model returns rgb, so we project them into coeffs (only SH supported)
             rgb, sigma = project_nerf_to_sh(nerf, args.sh_deg, points)
 
-        if tree.data_format.format == tree.data_format.RGBA:
+        if tree.data_format.format == tree.data_format.RGBA:  
             rgb = rgb.reshape(-1, args.samples_per_cell, tree.data_dim - 1);
             sigma = sigma.reshape(-1, args.samples_per_cell, 1);
             sigma_avg = sigma.mean(dim=1)
