--- ../../plenoxels/opt/util/config_util.py	2024-02-13 12:06:07.930166019 -0500
+++ ../google/plenoxels/opt/util/config_util.py	2024-02-13 10:46:24.452645156 -0500
@@ -5,7 +5,7 @@
 
 
 def define_common_args(parser : argparse.ArgumentParser):
-    parser.add_argument('data_dir', type=str)
+    parser.add_argument('--data_dir', type=str)
 
     parser.add_argument('--config', '-c',
                          type=str,
