--- ../../plenoctree/nerf_sh/nerf/utils.py	2024-02-13 12:39:49.531388387 -0500
+++ ../google/plenoctree/nerf_sh/nerf/utils.py	2024-02-13 10:46:24.356645265 -0500
@@ -236,7 +236,7 @@
         return
     pth = path.join(args.config + ".yaml")
     with open_file(pth, "r") as fin:
-        configs = yaml.load(fin, Loader=yaml.FullLoader)
+        configs = yaml.safe_load(fin)
     # Only allow args to be updated if they already exist.
     invalid_args = list(set(configs.keys()) - set(dir(args)))
     if invalid_args:
